zhaquirks/tuya/__init__.py

```diff
@@ -1532,35 +1532,38 @@ class AttributeDefs(BaseAttributeDefs):
         ),
     }
 
-    dp_to_attribute: dict[int, DPToAttributeMapping] = {}
+    dp_to_attribute: dict[int, DPToAttributeMapping | list[DPToAttributeMapping]] = {}
     data_point_handlers: dict[int, str] = {}
 
     def __init__(self, *args, **kwargs):
         """Initialize the cluster and mark attributes as valid on LocalDataClusters."""
         super().__init__(*args, **kwargs)
-        for dp_map in self.dp_to_attribute.values():
+        for dp_map in self._dp_to_attributes.values():
             # get the endpoint that is being mapped to
             endpoint = self.endpoint
-            if dp_map.endpoint_id:
-                endpoint = self.endpoint.device.endpoints.get(dp_map.endpoint_id)
-
-            # the endpoint to be mapped to might not actually exist within all quirks
-            if not endpoint:
-                continue
-
-            cluster = getattr(endpoint, dp_map.ep_attribute, None)
-            # the cluster to be mapped to might not actually exist within all quirks
-            if not cluster:
-                continue
-
-            # mark mapped to attribute as valid if existing and if on a LocalDataCluster
-            attr = cluster.attributes_by_name.get(dp_map.attribute_name)
-            if attr and isinstance(cluster, LocalDataCluster):
-                # _VALID_ATTRIBUTES is only a class variable, but as want to modify it
-                # per instance here, we need to create an instance variable first
-                if "_VALID_ATTRIBUTES" not in cluster.__dict__:
-                    cluster._VALID_ATTRIBUTES = set()
-                cluster._VALID_ATTRIBUTES.add(attr.id)
```

Comment:
Well, this code also never expected a tuple, so the "old multi DP code".
Needs updating in a future PR if we keep that code around.

Reply:
Personally I would prefer to drop the tuple from `DPToAttributeMapping`. Even if a few custom quirks break, I think it is worth it to have the code a bit clean. Except if there are a lot of custom quirks using it.

Reply:
Ok, we probably should remove it soon then.
I guess we could also convert the old tuple-style to the new `_dp_to_attributes` format with the list in `__init__` if we really wanted to. Then, we wouldn't need to care about handling it again in `_dp_2_attr_update` or for the "mark attributes as valid". I think that could be nice. Otherwise, we can just remove it soon.

Reply:
I am fine with both removing or converting it to `_dp_to_attributes`. Personally, I prefer to remove it just to keep the code simpler and easier to maintain.

But let me know which approach you prefer and I'll do that. Is it fine if we leave that to a follow-up PR? This one is not breaking it anyway.

Reply:
yeah, that's fine.